packages/driver/src/cy/commands/sessions.ts

```diff
@@ -0,0 +1,822 @@
+      }
+
+      // backup session command so we can set it as codeFrame location for errors later on
+      const sessionCommand = cy.state('current')
+
+      // stringfy determinitically if we were given an object
+      id = typeof id === 'string' ? id : stringifyStable(id)
+
+      if (options) {
+        if (!_.isObject(options)) {
+          throw new Error('cy.session optional third argument must be an object')
+        }
+
+        const validopts = {
+          'validate': 'function',
+        }
+
+        Object.keys(options).forEach((key) => {
+          const expectedType = validopts[key]
+
+          if (!expectedType) {
+            throw new Error(`unexpected option **${key}** passed to cy.session options`)
+          }
+
+          const actualType = typeof options[key]
+
+          if (actualType !== expectedType) {
+            throw new Error(`invalid option **${key}** passed to cy.session options. Expected **${expectedType}**, got ${actualType}`)
+          }
+        })
+      }
+
+      let existingSession = getActiveSession(id)
+
+      if (!setup) {
+        if (!existingSession || !currentTestRegisteredSessions.has(id)) {
+          $errUtils.throwErrByPath('sessions.session.not_found', { args: { id } })
+        }
+      } else {
+        const isUniqSessionDefinition = !existingSession || existingSession.setup.toString().trim() !== setup.toString().trim()
+
+        if (isUniqSessionDefinition) {
+          if (currentTestRegisteredSessions.has(id)) {
+            throw $errUtils.errByPath('sessions.session.duplicateId', { id: existingSession.id })
+          }
+
+          existingSession = sessions.defineSession({
+            id,
+            setup,
+            validate: options.validate,
+          })
+
+          currentTestRegisteredSessions.set(id, true)
+        }
+      }
+
+      const _log = Cypress.log({
+        name: 'session',
+        message: `${existingSession.id > 50 ? `${existingSession.id.substr(0, 47)}...` : existingSession.id}`,
+        groupStart: true,
+        snapshot: false,
+      })
+
+      const dataLog = Cypress.log({
+        name: 'session',
+        sessionInfo: getSessionDetails(existingSession),
+        message: `${existingSession.id > 50 ? `${existingSession.id.substr(0, 47)}...` : existingSession.id}`,
+      })
+
+      async function runsetup (existingSession) {
+        Cypress.log({
+          name: 'Create New Session',
+          state: 'passed',
+          event: true,
+          type: 'system',
+          message: ``,
+          groupStart: true,
+        })
+
+        if (!hadValidationError) {
+          _log.set({
+            renderProps: () => {
+              return {
+                indicator: 'successful',
+                message: `(new) ${_log.get().message}`,
+              }
+            },
+          })
+        }
+
+        await navigateAboutBlank()
+
+        cy.then(() => sessions.clearCurrentSessionData())
```

Comment:
what's the reason to use `cy.then` instead of regular Promises/async/await inside of this function? at first i thought you were doing this because you used other chainables inside of cy.session, but i don't see any chainables besides `cy.then`, so it seems unnecessary to me to use it

Reply:
fixed